Geometry_Engine/Query/GeometryHash.cs

```diff
@@ -0,0 +1,409 @@
+                    pointTranslationFactor = 0;
+
+                result.AddRange(obj.Vertices[i].ToDoubleArray(pointTranslationFactor + translationFactor));
+            }
+
+            return result.ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("Get the number of faces that are attached to each control point, " +
+            "and use that count as a translation factor for control points.")]
+        private static double[] GeometryHash(this Mesh3D obj, double translationFactor)
+        {
+            translationFactor += (int)TypeTranslationFactor.Mesh3D;
+
+            var dic = new Dictionary<int, int>();
+
+            for (int i = 0; i < obj.Faces.Count; i++)
+            {
+                foreach (var faceIndex in obj.Faces[i].FaceIndices())
+                {
+                    if (dic.ContainsKey(faceIndex))
+                        dic[faceIndex] += i;
+                    else
+                        dic[faceIndex] = i;
+                }
+            }
+
+            List<double> result = new List<double>();
+
+            for (int i = 0; i < obj.Vertices.Count; i++)
+            {
+                int pointTranslationFactor;
+                if (!dic.TryGetValue(i, out pointTranslationFactor))
+                    pointTranslationFactor = 0;
+
+                result.AddRange(obj.Vertices[i].ToDoubleArray(pointTranslationFactor + translationFactor));
+            }
+
+            return result.ToArray();
+        }
+
+
+        /***************************************************/
+        /****  Vector                                   ****/
+        /***************************************************/
+
+        [Description("The GeometryHash for a Point is simply an array of 3 numbers composed by the Point X, Y and Z coordinates.")]
+        private static double[] GeometryHash(this Point obj, double translationFactor)
+        {
+            return obj.ToDoubleArray(translationFactor);
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for a Plane is the GeometryHash of its Origin point translated by its Normal vector.")]
+        private static double[] GeometryHash(this Plane obj, double translationFactor)
+        {
+            translationFactor += (int)TypeTranslationFactor.Plane;
+
+            return obj.Origin.Translate(obj.Normal).ToDoubleArray(translationFactor);
+        }
+
+
+        /***************************************************/
+        /****  Other methods                            ****/
+        /***************************************************/
+
+        // Fallback
+        private static double[] GeometryHash(this object obj, double translationFactor)
+        {
+            BH.Engine.Base.Compute.RecordError($"Could not find a {nameof(GeometryHash)} method for type {obj.GetType().FullName}.");
```

Comment:
Potential future proof (in case a type of IGeometry is added outside the core, with its own implementation of GeometryHash in another Engine) is TryRunExtensionMethod here and return value if found, if not, return error.

Downside I guess is that that would be hashes outside our control, and unable to confirm that the implementation is valid, but also the case for other similar cases, such as IElement methods and similar.

Reply:
Why not! Sure I'll add that.